go/proxyd/methods.go

```diff
@@ -5,129 +5,262 @@ import (
+		}
 	}
-	return curBlock <= blockNum+numBlockConfirmations, nil
+
+	key := e.cacheKey(req)
+	return putImmutableRPCResponse(ctx, e.cache, key, req, res)
 }
 
-type EthGetBlockRangeMethod struct {
+type EthGetBlockRangeMethodHandler struct {
+	cache               Cache
 	getLatestBlockNumFn GetLatestBlockNumFn
 }
 
-func (e *EthGetBlockRangeMethod) CacheKey(req *RPCReq) string {
+func (e *EthGetBlockRangeMethodHandler) cacheKey(req *RPCReq) string {
 	start, end, includeTx, err := decodeGetBlockRangeParams(req.Params)
 	if err != nil {
 		return ""
 	}
 	return fmt.Sprintf("method:eth_getBlockRange:%s:%s:%t", start, end, includeTx)
 }
 
-func (e *EthGetBlockRangeMethod) IsCacheable(req *RPCReq) (bool, error) {
+func (e *EthGetBlockRangeMethodHandler) cacheable(req *RPCReq) (bool, error) {
 	start, end, _, err := decodeGetBlockRangeParams(req.Params)
 	if err != nil {
 		return false, err
 	}
 	return !isBlockDependentParam(start) && !isBlockDependentParam(end), nil
 }
 
-func (e *EthGetBlockRangeMethod) RequiresUnconfirmedBlocks(ctx context.Context, req *RPCReq) (bool, error) {
-	curBlock, err := e.getLatestBlockNumFn(ctx)
-	if err != nil {
-		return false, err
+func (e *EthGetBlockRangeMethodHandler) GetRPCMethod(ctx context.Context, req *RPCReq) (*RPCRes, error) {
+	if ok, err := e.cacheable(req); !ok || err != nil {
+		return nil, err
+	}
+
+	key := e.cacheKey(req)
+	return getImmutableRPCResponse(ctx, e.cache, key, req)
+}
+
+func (e *EthGetBlockRangeMethodHandler) PutRPCMethod(ctx context.Context, req *RPCReq, res *RPCRes, blockNumberSync uint64) error {
+	if ok, err := e.cacheable(req); !ok || err != nil {
+		return err
 	}
 
 	start, end, _, err := decodeGetBlockRangeParams(req.Params)
 	if err != nil {
-		return false, err
-	}
-	if isBlockDependentParam(start) || isBlockDependentParam(end) {
-		return true, nil
+		return err
 	}
-	if start == "earliest" && end == "earliest" {
-		return false, nil
+	curBlock, err := e.getLatestBlockNumFn(ctx)
+	if err != nil {
+		return err
 	}
-
 	if start != "earliest" {
 		startNum, err := decodeBlockInput(start)
 		if err != nil {
-			return false, err
+			return err
 		}
 		if curBlock <= startNum+numBlockConfirmations {
-			return true, nil
+			return nil
 		}
 	}
 	if end != "earliest" {
 		endNum, err := decodeBlockInput(end)
 		if err != nil {
-			return false, err
+			return err
 		}
 		if curBlock <= endNum+numBlockConfirmations {
-			return true, nil
+			return nil
 		}
 	}
-	return false, nil
+
+	key := e.cacheKey(req)
+	return putImmutableRPCResponse(ctx, e.cache, key, req, res)
+}
+
+type EthCallMethodHandler struct {
+	cache               Cache
+	getLatestBlockNumFn GetLatestBlockNumFn
+}
+
+func (e *EthCallMethodHandler) cacheKey(req *RPCReq) string {
+	type ethCallParams struct {
+		From     string `json:"from"`
+		To       string `json:"to"`
+		Gas      string `json:"gas"`
+		GasPrice string `json:"gasPrice"`
+		Value    string `json:"value"`
+		Data     string `json:"data"`
```

Comment:
You can see the data structure that geth uses for this here: https://github.com/ethereum/go-ethereum/blob/master/internal/ethapi/transaction_args.go
I think using strings is fine for our usecase but we will need to remember to update this once we support typed transactions

Reply:
how would this change with typed transactions? This structure is used only for eth_call.